### PR TITLE
Amazon S3 (New Pull request)

### PR DESCRIPTION
**Pull request by rmillo8**
Hi,
  I made one little changes (the config!): here's the new pull request on V2.

Let me know if that's ok.

Raf

### CODE REVIEW

File: build/NuSpecs/ImageProcessor.Web.Config.nuspec [hunk 1]
```diff
@@ -2,7 +2,11 @@
 <package xmlns="http://schemas.microsoft.com/packaging/2012/06/nuspec.xsd">
   <metadata>
     <id>ImageProcessor.Web.Config</id>
+<<<<<<< HEAD
```
Comment (JimBobSquarePants) [inline build/NuSpecs/ImageProcessor.Web.Config.nuspec]: Broken I'm afraid.

File: src/ImageProcessor.Web/Configuration/Resources/cache.config.transform [hunk 2]
```diff
@@ -1,9 +1,23 @@
...
+    <!-- The virtual path to the disk cache location. -->
+
+    <!--  <cache name="DiskCache" type="ImageProcessor.Web.Caching.DiskCache, ImageProcessor.Web" maxDays="365">
       <settings>
         <setting key="VirtualCachePath" value="~/app_data/cache"/>
       </settings>
+    </cache>-->
+    <cache name="AmazonS3Cache" type="ImageProcessor.Web.Plugins.AmazonS3Cache.AmazonS3Cache, ImageProcessor.Web.Plugins.AmazonS3Cache" maxDays="365">
+      <settings>
+        <setting key="awsAccessKey" value="" />
```
Comment (JimBobSquarePants) [inline src/ImageProcessor.Web/Configuration/Resources/cache.config.transform]: Can we have a format guide for these settings?

File: src/ImageProcessor.sln [hunk 3]
```diff
@@ -1,8 +1,6 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 2013
```
Comment (JimBobSquarePants) [inline src/ImageProcessor.sln]: What's going on here?

File: src/ImageProcessor.sln [hunk 4]
```diff
@@ -197,23 +200,36 @@ Global
 		{3C805E4C-D679-43F8-8C43-8909CDB4D4D7}.Release|Mixed Platforms.ActiveCfg = Release|Any CPU
 		{3C805E4C-D679-43F8-8C43-8909CDB4D4D7}.Release|Mixed Platforms.Build.0 = Release|Any CPU
 		{3C805E4C-D679-43F8-8C43-8909CDB4D4D7}.Release|x86.ActiveCfg = Release|Any CPU
+		{A4BA5609-9679-4FD4-A5B9-E3E791FBFFC5}.All|Any CPU.ActiveCfg = Release|Any CPU
```
Comment (JimBobSquarePants) [inline src/ImageProcessor.sln]: And here?

**Comment by JimBobSquarePants** [comment]
I'm afraid this pull request is broken. Are you working on v2 at your end and fetching the latest?

**Comment by rmillo8** [comment]
Let me try again later... (sorry!)

**Comment by JimBobSquarePants** [comment]
No worries; it happens :smile:

**Comment by JimBobSquarePants** [comment]
Hey there, 

I've manually added the code you've written to my solution since the merge was not going to happen easily. This means I'm going to close the two pull requests. Thanks for the code :+1: 

There's still a couple of thing in there that I need to get sorted before I can release.

- cache cleanup
- checking expired images from another bucket.

If you have any time and would like to help there that would be greatly appreciated.

**Comment by AliSheikhTaheri** [comment]
Hi There,

I know this is an old pull request, but has this code ever been merged to the master branch ? if not how can I use image processor in Umbraco that stores the cache file on Amazon S3?

Thanks
Ali

**Comment by JimBobSquarePants** [comment]
Hi there, 

We've jut finished some updates to the library and It's in the [Framework](https://github.com/JimBobSquarePants/ImageProcessor/tree/3611afc60096244314e00984eeddd5cf69cf92c5/src/Plugins/ImageProcessor.Web/ImageProcessor.Web.Plugins.AmazonS3Cache) branch just now. Doing some final tidy up to ensure it matches the date matching abilities of the Azure cache and it'll be good to release.

I'll do that asap.

Cheers

James

**Comment by AliSheikhTaheri** [comment]
Hi James,

That's a brilliant news, and thanks for your quick reply.

Do you have any time frame in mind as I am starting a project and I need to know when the final version will be ready?

Great work! 

Cheer
Ali

**Comment by JimBobSquarePants** [comment]
No worries :smile: 

I'll try and get it sometime tomorrow.

**Comment by AliSheikhTaheri** [comment]
Hi James, 

Just wanted to know if you got a chance to merge the pull request into master branch yet?

Thanks in advance 

Ali

**Comment by JimBobSquarePants** [comment]
Sorry, not yet. Had a disaster at work on Friday. I'll give it a pop tomorrow.

**Comment by AliSheikhTaheri** [comment]
James thanks for the update.

Look forward to hear the good news :)

**Comment by JimBobSquarePants** [comment]
@AliSheikhTaheri I'm gonna need a hand with this if you are able. 

There are a few strange things going on with the code and I don't really know the S3 SDK. Could you look at the class in the Framework branch for me? You'll see a bunch of `TODO:` comments that need sorting before I can release. 

https://github.com/JimBobSquarePants/ImageProcessor/blob/a21be06e912902ec4fedffd992939335e8c8405b/src/Plugins/ImageProcessor.Web/ImageProcessor.Web.Plugins.AmazonS3Cache/AmazonS3Cache.cs

**Comment by AliSheikhTaheri** [comment]
Hi @JimBobSquarePants , I asked my friend,@rmillo8, to have a look at it

**Comment by JimBobSquarePants** [comment]
Thanks! :+1:

**Comment by rmillo8** [comment]
Hi James,
 here are my thoughts on the "TODOs".

//TODO: line 69
`private string imageProcessorCachePrefix = @"imageprocessor_cache/";`
This is the name of the folder (in S3) where all cached files are stored.
E.g. \imageprocessor_cache\e\1\2\7\8\7/e127877845be1d9810b6a72a4e071679712d6112.png
I thought it wasn't necessary to have this customizable in cache.config

// TODO: line 134,line 348, line 377, line 391
I've been using S3 as a CDN, that's why I haven't made a general distinction between "cachedCDNRoot" and "cachedRoot".
If you wanted to allow users to store the cache in a CDN, you'd just need to add another "key" in cache.config and toggle between the two roots.

Let me know what you thinks.

Raf

**Comment by JimBobSquarePants** [comment]
Hi Raf,

Thanks for helping again.

What I'd like to do, as much as possible, is to make this plugin as consistent as possible with the Azure Blob Cache plugin.  

With 

``` c#
//TODO: line 69
private string imageProcessorCachePrefix = @"imageprocessor_cache/";
```

It's important to be able to set the prefix. There are situations where user will want to use different ones in different build environments (dev, staging, production) but a single bucket. This practise is quite common on Azure where a single storage account can have multiple blob containers.

``` c#
// TODO: line 134,line 348, line 377, line 391
```

In Azure there are two urls that you can use to request a cached image. The uncached, direct, url to the blob and the cached, distributed CDN url. I'm assuming that S3 does something very similar? At times when developing it can take time for the CDN to be fully initialized so it is important to be able to switch the CDN on and off. 

There is also the ability to request all images from the CDN and have the CDN redirect the request to the Blob url when the image hasn't been processed and cached. This can improve scaling due to the lack of 302 requests. Again I'm assuming S3 must do something very similar.

Does that make more sense?

Cheers

James

**Comment by electricsheep** [comment]
Really keen for this plug-in too!

Regarding a CDN with S3, the way to do this is by using CloudFront. It's not technically part of S3 - http://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/MigrateS3ToCloudFront.html

**Comment by rmillo8** [comment]
Hi James,
  I don't really agree with your first point re different cache- prefixes for different environments: in S3 you'd rather create different buckets for different environments as to ensure that content uploaded during development doesn't get mixed up with live contet. In this scenario you'd create separate cache.config files with different keys, secrets and bucket names.
But, as I said, it can be added to the cache.config with no problem.

Regarding S3 and CDN: with AWS you can serve content in your S3 bucket using CloudFront (thanks @electricsheep ), but I wouldn't call this an S3 feature.
Normally I would manage CDN distributions outside, say, ImageProcessor configuration as I wouldn't want to have to change several different things in several different places if I were to switch (or switch off) my CDN. But as I said, we can easily amend this and allow users to add their own CDNroot if needed. Code will mock AzurePlugin code, as there won't be anything specific to S3.

I'll send another pull request in the next few days if you want.

Raf
